File: apps/web/lib/forward-picker-model.ts
/**
 * Machine à états du picker de transfert hybride (spec 2026-08-19, Volet A.6/D) :
 * bouton par-ligne = envoi immédiat, tap ligne = multi-sélection, une cible déjà
 * servie n'est plus sélectionnable, aucun doublon possible par construction.
 * RÈGLE JUMELLE : apps/ios/Meeshy/Features/Main/Components/ForwardPickerModel.swift
 * — toute évolution touche les deux.
 */

export type TargetState = 'idle' | 'selected' | 'sending' | 'sent' | { failed: string };

export class ForwardPickerModel {
  private states: ReadonlyMap<string, TargetState> = new Map();

  state(id: string): TargetState {
    return this.states.get(id) ?? 'idle';
  }

  selectedIds(): string[] {
    return [...this.states.entries()]
      .filter(([, state]) => state === 'selected')
      .map(([id]) => id)
      .sort();
  }

  tapRow(id: string): void {
    const current = this.state(id);
    if (current === 'selected') {
      this.setState(id, 'idle');
      return;
    }
    if (current === 'sending' || current === 'sent') return;
    // `idle` ET `{failed}` deviennent sélectionnables : une cible en échec doit
    // pouvoir rejoindre un envoi groupé (parité iOS `case .idle, .failed`).
    this.setState(id, 'selected');
  }

  beginSend(id: string): boolean {
    const current = this.state(id);
    if (current === 'sent' || current === 'sending') return false;
    this.setState(id, 'sending');
    return true;
  }

  finishSend(id: string, ok: boolean, reason?: string): void {
    if (this.state(id) !== 'sending') return;
    this.setState(id, ok ? 'sent' : { failed: reason ?? '' });
  }

  beginBatch(): string[] {
    const batch = this.selectedIds();
    batch.forEach((id) => this.setState(id, 'sending'));
    return batch;
  }

  private setState(id: string, state: TargetState): void {
    this.states = new Map([...this.states.entries(), [id, state]]);
  }
}

File: apps/ios/Meeshy/Features/Main/Components/ForwardPickerModel.swift
import Foundation

/// Machine à états PURE du picker de transfert — la sémantique hybride exacte
/// décidée par le user (spec 2026-08-19, Volet A.6) :
/// 1. le bouton par-ligne envoie IMMÉDIATEMENT à cette seule cible ;
/// 2. toucher la ligne sélectionne (mode multi) — un bouton global envoie aux
///    sélectionnées ;
/// 3. une cible déjà servie n'est plus sélectionnable ;
/// 4. envoyer par-ligne une cible sélectionnée la retire de la sélection —
///    l'envoi groupé ne peut JAMAIS produire un doublon, par construction.
/// RÈGLE JUMELLE : apps/web/lib/forward-picker-model.ts — toute évolution
/// touche les deux sites. Ce fichier est aussi compilé dans
/// `MeeshyShareExtension` : il n'importe QUE `Foundation`, et
/// `ForwardPickerModelPortabilityGuardTests` l'y maintient.
struct ForwardPickerModel: Equatable {
    enum TargetState: Equatable {
        case idle
        case selected
        case sending
        case sent
        case failed(String)
    }

    private(set) var states: [String: TargetState] = [:]

    func state(of id: String) -> TargetState { states[id] ?? .idle }

    var selectedIds: [String] {
        states.filter { $0.value == .selected }.map(\.key).sorted()
    }

    var hasSelection: Bool { states.contains { $0.value == .selected } }

    mutating func tapRow(_ id: String) {
        switch state(of: id) {
        case .idle, .failed: states[id] = .selected
        case .selected: states[id] = .idle
        case .sending, .sent: break
        }
    }

    @discardableResult
    mutating func beginSend(_ id: String) -> Bool {
        switch state(of: id) {
        case .idle, .selected, .failed:
            states[id] = .sending
            return true
        case .sending, .sent:
            return false
        }
    }

    /// L'issue est PRIMITIVE (`succeeded` + `reason`) et non l'issue riche
    /// déclarée dans `MessageForwardService.swift` : ce fichier est compilé
    /// DANS `MeeshyShareExtension`, qui n'a aucune dépendance SDK, alors que
    /// `MessageForwardService.swift` `import MeeshySDK`. La jumelle web a la
    /// même signature depuis toujours (`forward-picker-model.ts:44` —
    /// `finishSend(id, ok, reason?)`). La traduction depuis l'issue riche
    /// appartient à l'app (`succeeded` / `.failureReason`, portés par une
    /// extension côté service).
    mutating func finishSend(_ id: String, succeeded: Bool, reason: String?) {
        guard state(of: id) == .sending else { return }
        states[id] = succeeded ? .sent : .failed(reason ?? "")
    }

    mutating func beginBatch() -> [String] {
        let ids = selectedIds
        for id in ids { states[id] = .sending }
        return ids
    }
}
